### ISSUE TITLE
Application testing improvements for Meteor 1.3

### ISSUE THREAD

**Issue by benjamn** [meteor/meteor#5784]
- [x] Find tests in `tests/` subdirectories
- [x] Optimize `meteor test-app` - copy build dir before starting.
- [x] `Meteor.isTest`, `Meteor.isIntegrationTest`, `Meteor.isUnitTest`.
- [x] Implement `meteor test-app --unit`
- [x] `meteor test-app` to find all test files, even inside `imports/`
- [x] Why do integration tests no longer correctly hide the app UI?
- [x] Why do Blaze unit tests fail on the Todos app
- [x] Can we actually have two test reporter packages that both expose Mocha added to the app?
- [x] Simplify process for building test runner packages (clear CSS, clear DOM, stop router).
- [x] Require passing in a driver package option
- [x] Figure out a directory scheme for acceptance (i.e. outside of Meteor) tests #6321
- [x] Add help text so that `curmeteor help test` works
- [x] Add test to skeleton and example apps. #6303 #6306

**Comment by philipmjohnson** [comment]
Hallelujah!  Relief from app-level testing madness is finally in site!

**Comment by Sanjo** [comment]
Let me know if you have something to show or want some input. I did a lot of the work in Velocity, including the meteor --test-app implementation in the Velocity Meteor fork.

**Comment by avital** [comment]
Here are my current thoughts:

The `meteor test-app` command (that will later be renamed to plain `meteor test`) will be modeled after `meteor test-packages`. Here's how it will work:

* Instead of Velocity mirrors that basically make copies of the app, it will bundle the same app directory into a new temporary build directory. It still uses the same "watch files for changes" logic (which can run in parallel to running the app normally), so you get hot code push for tests.

* This process adds a test reporter package that can be specified as a command line argument as part of this process. This doesn't affect any parallel non-test runs of the app.

* Running with `--unit` only loads test files explicitly passed in as arguments and their dependencies. Running with `--integration` also loads the entire app.

This isn't the entire story but I thought I'd jot down these thoughts in parallel to working on them.

@Sanjo -- what do you think? Have I correctly understood the purpose of Velocity mirrors? Will this plan resolve the need for using them, at least for the case of not running parallel test chunks on different servers?

**Comment by stubailo** [comment]
You mentioned a test reporter package, what about the test runner? For example mocha, tinytest, etc. Right now test-packages takes a special argument, will test-app have something similar?

**Comment by avital** [comment]
@stubailo If I understand correctly, letting you orthogonally choose a test runner package and a test reporter package requires that Meteor also have a standard for reporting test results to the server. I believe that's part of what Velocity does.

I was hoping to not have to do that just right now, and instead say that you choose a package that bundles both reporter and test framework.

This is basically the same as what exists in Node, where if you use mocha you can choose one of the mocha reporters. You can't directly use a mocha reporter with say, tape. (Incidentally, looks like there's a thing called [tap](http://testanything.org/) that is an attempt at being a general purpose "test results" protocol/format).

Also looks like at least in some circles [Mocha is the clear dominant testing framework](https://twitter.com/mikeal/status/680834926334181376).

I would rather us basically say, "write tests using Mocha" and let innovation happen on the test reporter front. I don't really see what we have to gain by letting people innovate with testing libraries. I could see the downside -- fracturing the community.

What do you think?

**Comment by stubailo** [comment]
It sounds like you're saying that "test reporter" in your comment above also includes "test runner", which is totally fine. I don't think you need to be able to pick them independently, I just didn't see anything about mocha/tinytest/test runners in your original comment.

I'm totally OK just supporting an `abc:mocha` package that includes mocha itself, the test result management, and the UI for displaying the results.

**Comment by avital** [comment]
Yeah. That was my suggestion. I'll update the comment.

**Comment by Sanjo** [comment]
> Instead of Velocity mirrors that basically make copies of the app, it will bundle the same app directory into a new temporary build directory. It still uses the same "watch files for changes" logic (which can run in parallel to running the app normally), so you get hot code push for tests.

Velocity already bundles the same app directory plus some specified tests into a new temporary build directory. It also works very similar to how meteor test-packages work. You can find the implementation here: https://github.com/meteor-velocity/meteor/blob/velocity/tools/cli/commands.js#L304-L345.

I think the major performance loss is that a separate Meteor build tool instance is used for the mirror. 
There is also a speed problem for unit tests because currently the whole app needs to be built in order to run the unit tests. With the current implementation in Velocity that's too slow for TDD. I think what you suggest with `--unit`, handles this.

> This process adds a package that includes a test-reporter and a testing framework. That package can be specified as a command line argument as part of this process. This doesn't affect any parallel non-test runs of the app.

I think that's fine to start with. From the experience with Velocity I think the benefits of having a unified test result format weren't that big. Some people prefer CLI reporting over GUI reporting. GUI reporting allows a richer UI. So having the option to choose a reporter is still valuable. But that can probably just be an environment variable that is handled by the testing package.

> Running with --unit only loads test files explicitly passed in as arguments and their dependencies. Running with --integration also loads the entire app.

Can you also run all unit tests with the --unit mode?

How do I run all my unit and integration tests (what commands)?

**Comment by pauldowman** [comment]
@avital [This poll](https://ponyfoo.com/articles/javascript-developer-survey-results#what-unit-testing-libraries-do-you-use) might be a bit more representative than the Twitter poll you linked above. E.g. the Twitter poll didn't even include Jasmine, which is still popular. (Although I don't think Karma should be in there, I guess testing framework vs test runner is confusing).

**Comment by jiku** [comment]
Maybe worth noting that [Mocha 3](https://github.com/mochajs/mocha/issues/1969) might be [borrowing](https://github.com/sindresorhus/ava/issues/145) a lot from AVA, and that the latter, while exiting and better in many ways, [probably won't](https://twitter.com/sindresorhus/status/669200896589299712) be as widely used.

**Comment by avital** [comment]
@Sanjo Thanks for pointing me to the Velocity fork of Meteor. That's pretty much exactly what I was planning to do, including modifying `ProjectContext` to let you specific a directory to use instead of your app's `.meteor/local`.

Sorry for the basic question but what is the fork of Velocity used for? Do you have to use it if you want to use Velocity in your app?

In my proposal we'd still have a separate build process for tests (in fact, a separate command -- you'd run `meteor test --port XXX` explicitly). But if your machine has multiple cores that should be just as fast as having one build process running, no?

As to your question about how to specify which tests to run, I was thinking that if Meteor is moving towards the direction of more explicitly including files, then there shouldn't be a general notion of "all test files"; instead there'd be files like `foo_tests.js` that `imports` all the files that define tests that test the feature "foo" in your app. And you could pass globs to `meteor test` that search for say all `*tests.js` files if you really want to.

I don't think we will let people run unit and integration tests as part of the same run, since that'd require building the app in a different way. You could, say, run `meteor test --unit *unit_tests.js -p 5000` and `meteor test --integration *integration_tests.js -p 6000` in another shell.

Curious to hear what you think.

**Comment by avital** [comment]
Another related question I'm not sure about: If you simply run your app does it bundle the tests with it, but not run them? Or does it somehow know how to exclude test files?

If it's the former, you'd have to add the testing framework package to your app at all times, not just when running tests.

If it's the latter, we'd have to ensure a convention that tests live in tests/ subdirectories  or are always named test_*. Or require that they are always located inside an `imports/` directory so that they get evaluated lazily?

**Comment by Sanjo** [comment]
> Sorry for the basic question but what is the fork of Velocity used for? Do you have to use it if you want to use Velocity in your app?

The Velocity mirrors run with the fork. For build and including a set of tests with the mirror.

> In my proposal we'd still have a separate build process for tests (in fact, a separate command -- you'd run meteor test --port XXX explicitly). But if your machine has multiple cores that should be just as fast as having one build process running, no?

I think the hard disk I/O is also a bigger factor. Also with my usage I usually ran 2 to 3 mirrors. One for server integration tests, one for client integration tests, one for end to end / functional tests. But it's also that the Meteor build tool is currently not that fast in general for larger apps. I haven't tested the new speed improvements beyond 1.2.

> As to your question about how to specify which tests to run, I was thinking that if Meteor is moving towards the direction of more explicitly including files, then there shouldn't be a general notion of "all test files"; instead there'd be files like foo_tests.js that imports all the files that define tests that test the feature "foo" in your app. And you could pass globs to meteor test that search for say all *tests.js files if you really want to.

Glob patterns for test files is how the most test runners do it and it works well. The test runners Karma and Wallaby treat each test file as entry point. That works well for unit testing. I think having to import all unit test files from one file is a bad idea because unit test files should be independent. And I think it's a good practice to put them next to the source code files. So you might want to introduce an extension for them. E.g. .spec.js and .tests.js.

> I don't think we will let people run unit and integration tests as part of the same run, since that'd require building the app in a different way. You could, say, run meteor test --unit *unit_tests.js -p 5000 and meteor test --integration *integration_tests.js -p 6000 in another shell.

Jup, makes sense to me. I think a good goal for unit testing would be that hundreds of unit tests can run in under two seconds. That would be really awesome for TDD.

> Another related question I'm not sure about: If you simply run your app does it bundle the tests with it, but not run them? Or does it somehow know how to exclude test files?

I think it would make sense to treat every test file as entry point, when your build the app for --unit or --integration. So the test files would get loaded after the Meteor packages and app code for integration testing. That's how I do it with my latest Wallaby config and it works well.
When you just start your app, tests should not be included, like it is today with the /tests folder. But I would extend this to files the .spec.js or .tests.js extension.

**Comment by raix** [comment]
Requirements for the test suite:
1. Fast startup (avoid starting the meteor server)
2. Reliable
3. Fast when doing TDD (eg. using ffdescribe or describe.only etc my limit is 50-500ms)
4. Fast CI support CI on multiple nodes
5. Debugable
6. Code coverage
7. Support upcomming tools like wallaby etc.

-

We currently have to decouple meteor in order to leverage existing tools like mocha/karma/istanbul - if meteor core packages existed as npm packages that might change things?

For new users configuring package.json to work with multiple tests suites/babel etc. and adding configurations for it all is really just boilerplating, it might be a blocker for newcomers, maybe thats were meteor can assist eg. help generate the boiler-plating.

I think it's worthwhile to avoid adding extra layers if not needed, it makes it harder to debug/maintain/unstable/slow. *(Requirement > Nice to have, Standards > Custom)*

**Comment by tmeasday** [comment]
Some work has been done towards these goals, read about it here: https://forums.meteor.com/t/a-first-implementation-of-meteor-app-testing-for-1-3/17097

**Comment by Sanjo** [comment]
@avital Just FYI, to make Velocity compatible with Meteor 1.3, we will need to merge Meteor 1.3 and the Velocity Meteor fork changes from the pull request: https://github.com/meteor/meteor/pull/4353.
Those changes are only necessary for the mirrors. So maybe we can make Velocity work the new mechanism that you are implementing for integration testing.

**Comment by avital** [comment]
@Sanjo Sorry, I don't fully understand what you're asking for. Can you elaborate?

**Comment by avital** [comment]
Like, what has changed with Meteor 1.3 that makes us need to merge that pull request? And really, what is left in Velocity after we finish this testing feature in 1.3? Will people still use Velocity? For what purpose?

**Comment by raix** [comment]
+1 for plain js test suites - keep it simple, don't add more layers to debug/maintain.

**Comment by Sanjo** [comment]
I think I have read somewhere in the forums that Velocity can still be used in Meteor 1.3.
When all actively maintained and used testing packages that currently use Velocity have migrated to the new Meteor 1.3 testing primitives, then there will be no need for Velocity anymore.

So I'm with you. Just remove the Velocity support by removing the run-velocity.js and `meteor --test` (velocity command) from Meteor core. Because Velocity will not work without me releasing a new version of the Velocity Meteor fork and updating velocity:core.
I can inform all Velocity package testing authors and then they can act or not.

**Comment by tmeasday** [comment]
@Sanjo, it would be good to get an idea of what the list of "actively maintained and used testing packages" are. Do you have a good grasp on it.

I think we should figure out if we want 1.3 to launch with:

1. No support for Velocity, just the new `meteor test-app` command. This would require
  a. All important packages to be updated, or to update soon after
  b. Some kind of migration guide

2. Support for both velocity (via the fork) and the command. So people could run both types of tests as they migrate to the new system.

I've been pondering what the migration guide would look like. I think it might be as simple as:

1. Rename some test files.
2. Update / add / remove some key testing packages.
3. Run app tests in this new way.

Do you think it will be as simple as that, or is there complexity that I am missing? Is there a good canonical example of an app built with some velocity tests that I can try the above process on?

**Comment by Sanjo** [comment]
There are four Velocity testing packages:

* [sanjo:jasmine](https://github.com/Sanjo/meteor-jasmine) - I will update the package to use the new Meteor 1.3 testing APIs
* [xolvio:cucumber](https://github.com/xolvio/meteor-cucumber) - Deprecated in favor of using Chimp directly
* [mike:mocha](https://github.com/mad-eye/meteor-mocha-web)
* [rsbatech:robotframework](https://github.com/rjsmith/meteor-robotframework)

> I've been pondering what the migration guide would look like. I think it might be as simple as:

> Rename some test files.
> Update / add / remove some key testing packages.
> Run app tests in this new way.

> Do you think it will be as simple as that, or is there complexity that I am missing?

Yes, I think so, when the used testing framework is supported for Meteor 1.3. So there will be no need to rewriting tests.

> Is there a good canonical example of an app built with some velocity tests that I can try the above process on?

Yes, here: https://github.com/meteor-velocity/velocity-examples.

**Comment by timbotnik** [comment]
Following on @raix's # 6 above - it would be great to have some support for obtaining code coverage metrics along with tests.  Seems this is a useful metric to lots of teams, and would be available to many people if using vanilla node, ruby, or golang.  At first glance, it would seem non-trivial to get this going with Meteor, as you need to measure against the right set of code (taking into account client vs. server, the build pipeline, transpilation, etc) and probably install a number of dependencies and get them to work correctly together.  

I found a [forum post](https://forums.meteor.com/t/package-test-coverage-using-istanbul/14420) referencing an integration with `istanbul` which might make for a decent starting point.  Also [this npm package](https://www.npmjs.com/package/mocha-phantom-coverage-reporter) seems to have a solution for node + browser that combines phantom + htmlcov + mocha-lcov-reporter.

Curious what you guys think of the degree of difficulty here, as well as level of demand for a code coverage feature.

**Comment by tmeasday** [comment]
@timbotnik I think this is out of scope for this iteration. 

I'd say this is similar to another feature that the build tool enables us to make seamless for people: Jest-style auto-mocking for `import`s from within the module under test in unit test mode. That would be high on my personal list of testing features.

**Comment by avital** [comment]
We're not going to expand scope for this feature at this point.

**Comment by Sanjo** [comment]
One note on client side testing. I use [Karma Test Runner](https://karma-runner.github.io/0.13/index.html) to run the tests. It would be great if the Meteor build tool could compile the tests with source maps and I could then load the compiled tests in the Karma Test Runner.

Related:
* Tests compilation pull-request: https://github.com/meteor/meteor/pull/5457
* Use case issue: https://github.com/xolvio/meteor-jasmine/issues/326

**Comment by Urigo** [comment]
@rbabayoff how come you are not a part of this conversation with your [practicalmeteor:mocha](https://github.com/practicalmeteor/meteor-mocha) package?

**Comment by vangorra** [comment]
+1 @Sanjo. Meteor should handle the compilation of the test files, it already knows how to and understand the dev specific options. Ultimately it would make testing much easier.

**Comment by avital** [comment]
Hi @Sanjo @vangorra (and any other Karma fans)

I haven't used Karma before, but it looks interesting. We could consider working to ease use of Karma with Meteor but with my knowledge, I can't make that call. If you think this is important, can you write up a design document that explains:

1. How a workflow with Karma and Meteor be better than running tests with the `meteor test-app` command as [designed in the guide](https://github.com/meteor/guide/blob/458745803e596ce241190c23192925fd666c4ebb/content/testing.md)
2. What Meteor would have to change to make it work well with Karma

Much appreciated

**Comment by vangorra** [comment]
@avital,
In short, we should be able to write test-specs in whatever language meteor can support via compiler packages. The current implementation (using karma) is far to restrictive to make it work well.

Long:
The request here is that Meteor 1.3's test system should use Meteor's own code compilers to compile test specs themselves. Essentially Meteor would be handling test spec files exactly as it does with app files. So, for example, tests/my-test.ts would be transpiled to .meteor/local/build/.../tests/my-test.js, at which point the test sub-system would execute the transpiled file normally. This consideration would vastly improve the test creation process. In addition, applying post-compile processing becomes simpler as one only has to worry about an es5 file in the build dir. This possibly paves the way to supporting code-coverage tools too.

The current mode of operation requires creating a Velocity mirror that uses Karma. Then Karma is configured to process the source files. After it transpiles the source files, then Karma can execute the tests. While this works for es5 and coffeescript quite easily, this configuration fails when working with es6 (babel) or typescript. Both of which often can depend on a config file. Getting the aforementioned language to work required a lot of duplicate config handling that Meteor is already capable of handling.

**Comment by avital** [comment]
@vangorra I think the `meteor test-app` command as [described here](https://github.com/meteor/guide/blob/458745803e596ce241190c23192925fd666c4ebb/content/testing.md) does what you are suggesting (runs Meteor compiler plugins on all files including test files). Is this sufficient? The app is built into a different temporary subdirectory though (are you looking for direct access into `.meteor/build`, or do you just want the test app to run?).

I don't understand some details I think -- how would you use Karma's file watching in combination with Meteor's?

I really would like a proper document answering the questions I asked above. Treat this like we would internally -- write a document, like a design doc or a "request for comments". 

Or maybe the `app-testing` branch isn't mature enough for that, in which case I'll update on this thread at the point where I think that branch is useable enough to try exploring new features on.

**Comment by vangorra** [comment]
@avital The "testing guide" you provided clears up a lot of confusion. The design as is currently described is sufficient to fit my original request. 

What I like about the document:
- Tests exist next to the app code using a .test.* suffix. This really helps puny humans to easily keep track of what tests apply to which file.
- When run in test mode, the code is built into a different sub-directory. This strategy is similar to how Velocity makes mirrors to keep built code isolated and not interfere with development. It also opens the built code up to post-processing by code coverage tools.
- The strict separation of integration testing and acceptance testing. As in, integration testing effectively eliminates the UI. 
- Different testing runners like mocha can be used.

What wasn't particularly clear:
- When ```meteor test-app``` is ran, is this in a continuous dev mode? As in, when you modify a file, does the test automatically re-run? 
- There was mention of using selenium for acceptance testing. This is a good start but a more generic solution would be to allow package-based test targets. So a developer would create a selenium-test-target package which meteor can use to to run the tests. Installing multiple test-target packages would allow meteor to run all the client side tests on multiple platforms. This becomes incredibly useful for testing apps using saucelabs, phantomjs, etc. Thus enabling the dev to testing their app across multiple platforms and browsers.

Ultimately, the current design for ```meteor test-app``` is a good solution and solves a lot of problems faced by Velocity. In fact, this design renders the use of Karma and possibly Velocity as redundant.

**Comment by Sanjo** [comment]
I think this feature request is out of the scope for this issue and should be moved into a separate one, if considered to be implemented in the future.

> How a workflow with Karma and Meteor be better than running tests with the meteor test-app command as designed in the guide

Karma is a sophisticated client side test runner that has been developed on [for four years](https://github.com/karma-runner/karma/graphs/contributors). It allows you to run client unit test __fast__ in __real browsers__ and also works very well in a __continuous integration__ environment. It __supports all the widely used testing frameworks__ like Mocha and Jasmine.

> What Meteor would have to change to make it work well with Karma

Test runners like Karma and Wallaby support compile plugins, to compile the source files before executing them. If Meteor would expose a public API that can be used in node.js to compile source files, a plugin could be written for Karma and Wallaby. This would ensure that your source code has the same form when running your tests with Karma or Wallaby.

Currently as a workaround I use Webpack to compile the source code for Karma and Wallaby, because both have already a plugin for Webpack. That works for me.

**Comment by avital** [comment]
@vangorra Yes, `meteor test-app` will re-run your app when tests or your app code changes. Can't say more about Selenium. I think @tmeasday has been thinking more about that.

**Comment by tmeasday** [comment]
Right now we're treating acceptance testing more as a blackbox -- you just run the app in development mode as normal (you might add some fixture generating hooks into your development app for convenience), then point whatever tool you like at it. 

I think this is an example of something where a Meteor package isn't really required to ship this kind of thing to users; we already have the ability to do so easily in npm packages right?

**Comment by tmeasday** [comment]
@avital: `Meteor.isIntegrationTest` appears to be set when you `meteor run`.

**Comment by tmeasday** [comment]
@avital I'm seeing an error come out of `avital:mocha` when i run Todos as normal (todos/master)

**Comment by avital** [comment]
@tmeasday I am not seeing any errors (I tried both with the `release-1.3` branch and the `1.3-beta.11` release). I wonder what I'm doing differently?

**Comment by avital** [comment]
We forgot to announce here but the recent `1.3-beta.11` (including modules, testing and Cordova improvements) now includes support for unit and integration app testing (but only with mocha for now). Learn more about the release, try it out and give us feedback! https://github.com/meteor/meteor/issues/6266

**Comment by vjau** [comment]
About the guide here : https://github.com/meteor/guide/blob/testing-modules-content/content/testing.md
1°) The mocha to add is avital:mocha and not practicalmeteor:mocha
2°) The test command to run is "meteor test-app" and not "meteor test", which doesn't exist atm.
3°) practicalmeteor:mocha-web-reporter doesn't exist.

**Comment by stubailo** [comment]
@vjau I think that document is more aspirational than concrete at the moment.

**Comment by vjau** [comment]
I don't know if it's aspirational but for me (and probably for some of the 90 people that were pointed to it from the forum thread) it's rather confusing !

**Comment by tmeasday** [comment]
Thanks @vjau - you are right, we should have tightened this up a little more before pointing you at it. Thanks for trying it out ;)

**Comment by tmeasday** [comment]
@avital My spec for test file naming:

### Unit tests:
 - outside of a top level `tests/` directory, i.e. matching `test[s].*`, `*.test[s].*` and `otherpath/tests/`

### Integration tests:
 - inside top level `tests/` directory (with exception of below)

### Acceptance tests (ignored by all modes of meteor)
 - inside top level `tests/acceptance` directory.

What do you think?

**Comment by tmeasday** [comment]
@avital good news; this worked: https://github.com/practicalmeteor/meteor-mocha-console-runner/compare/master...tmeasday:master

**Comment by ubald** [comment]
When running app-tests it seems client and server tests run at the same time. Maybe I'm the one doing something weird but I end up with conflicting results when the client and the server use the same collections at the same time.
